common/djangoapps/student/roles.py

```diff
@@ -100,17 +145,29 @@ def get_roles(role):
         """
         return ACCESS_ROLES_INHERITANCE.get(role, set()) | {role}
 
+    # TODO: test that this works as expected
+    def get_all_roles_set(self):
+        return self._roles
+
+    def get_roles_by_course_id(self):
+        return self._roles_by_course_id
+
     def has_role(self, role, course_id, org):
         """
         Return whether this RoleCache contains a role with the specified role
         or a role that inherits from the specified role, course_id and org.
         """
-        return any(
+        res = False
+        course_id_string = get_role_cache_course_key(course_id)
+        course_roles = self._roles_by_course_id.get(course_id_string)
+        if not course_roles:
+            return False
+        res = any(
             access_role.role in self.get_roles(role) and
-            access_role.course_id == course_id and
             access_role.org == org
-            for access_role in self._roles
+            for access_role in course_roles
         )
```

Comment:
The PR description mentions that we are going from O(total_num_libs * accessible_libs_and_courses) down to O(total_num_libs). Am I correct this is precisely where that optimization is happening?

In other words: is the optimization that `self._roles` contained _all_ of a user's access roles across the whole system, whereas `course_roles` contains only their set of roles in the course in question?

Reply:
Yes, this is precisely where the optimization is happening.

> is the optimization that self._roles contained all of a user's access roles across the whole system, whereas course_roles contains only their set of roles in the course in question?

Yes. Let's understand the fundamental problem with the prior implementation: the previous code was iterating over every single role a user has, and some user have thousands of roles, because each library or course results in a new role. Now running a find operation here by iterating over those thousands of roles is very inefficient.
On top of that the code itself seems already slow, which I still don't completely understand, but that's fine - iterating over thousands of roles every time this function is called (and it's called for every library that exists on the platform) makes this much worse.
Just picking the key for the course or library you want to check from a dictionary instead of a check is a quick O(1) operation instead of an iteration over the whole set.
Now the reason why the optimization should work very well is that a user usually has at most a handful of roles for a course. So as long as this is not iterating over the ungrouped set it should significantly increase performance of the python code.

Reply:
Basically as pseudocode:
roles = dict[course_id]
is much quicker than
roles = iterate_over_set(all_roles).